### ISSUE TITLE
Support for Kafka versions 2.0.0+

### ISSUE THREAD

**Issue by arjun289** [kafkaex/kafka_ex#415]
I am trying to use `kafka_ex` with the latest version of Kafka, `2.6.0`.

I looked at the docker-compose file in the project which makes use of 0.11.0.1 which runs perfectly fine. 
However,  there are certain changes in the way configurations need to be done for running the latest versions of Kafka.

The `docker-compose.yml` with the latest configuration changes.
```
version: '3.2'

services:
  zookeeper:
    image: wurstmeister/zookeeper:3.4.6
    ports:
      - "2181:2181"
  kafka1:
    image: wurstmeister/kafka:latest
    ports:
      - "9093:9093"
    depends_on:
      - zookeeper
    volumes:
      - ./ssl:/ssl
    env_file: docker-compose-kafka.env
    environment:
      KAFKA_BROKER_ID: 1
      KAFKA_LISTENERS: INSIDE://kafka1:29093, OUTSIDE://localhost:9093
      KAFKA_ADVERTISED_LISTENERS: OUTSIDE://localhost:9093, INSIDE://kafka1:29093

  kafka2:
    image: wurstmeister/kafka:latest
    ports:
      - "9094:9094"
    depends_on:
      - zookeeper
    volumes:
      - ./ssl:/ssl
    env_file: docker-compose-kafka.env
    environment:
      KAFKA_BROKER_ID: 2
      KAFKA_LISTENERS: INSIDE://kafka2:29094, OUTSIDE://localhost:9094
      KAFKA_ADVERTISED_LISTENERS: INSIDE://kafka2:29094, OUTSIDE://localhost:9094

  kafka3:
    image: wurstmeister/kafka
    ports:
      - "9095:9095"
    depends_on:
      - zookeeper
    volumes:
      - ./ssl:/ssl
    env_file: docker-compose-kafka.env
    environment:
      KAFKA_BROKER_ID: 3
      KAFKA_LISTENERS: INSIDE://kafka3:29095, OUTSIDE://localhost:9095
      KAFKA_ADVERTISED_LISTENERS: INSIDE://kafka3:29095, OUTSIDE://localhost:9095

```

Docker env file 
```
#################################
# Common Kafka config
#################################

# Note: any property `something.bla` can be configured by setting `KAFKA_SOMETHING_BLA`.

######## zookeeper

KAFKA_ZOOKEEPER_CONNECT=zookeeper:2181
KAFKA_ZOOKEEPER_CONNECTION_TIMEOUT_MS=6000

######## advertised hosts names and protocols

# alternative to KAFKA_ADVERTISED_HOST_NAME is: HOSTNAME_COMMAND: ip addr | grep -o "inet [0-9.]*" | grep -v "127\.0\.0\.1"  | grep -o "[0-9.]*"
KAFKA_LISTENER_SECURITY_PROTOCOL_MAP=INSIDE:PLAINTEXT,OUTSIDE:PLAINTEXT
KAFKA_INTER_BROKER_LISTENER_NAME=INSIDE
######## SSL

# KAFKA_SSL_KEYSTORE_LOCATION=/ssl/server.keystore.jks
# KAFKA_SSL_KEYSTORE_PASSWORD=kafka_ex
# KAFKA_SSL_KEY_PASSWORD=kafka_ex
# KAFKA_SSL_TRUSTSTORE_LOCATION=/ssl/server.truststore.jks
# KAFKA_SSL_TRUSTSTORE_PASSWORD=kafka_ex
# KAFKA_SSL_SECURE_RANDOM_IMPLEMENTATION=SHA1PRNG

######## Config

KAFKA_DELETE_TOPIC_ENABLE=true
# KAFKA_NUM_NETWORK_THREADS=3
# KAFKA_NUM_IO_THREADS=8
# KAFKA_SOCKET_SEND_BUFFER_BYTES=102400
# KAFKA_SOCKET_RECEIVE_BUFFER_BYTES=102400
# KAFKA_SOCKET_REQUEST_MAX_BYTES=104857600
KAFKA_LOG_DIRS=/tmp/kafka_log
KAFKA_NUM_PARTITIONS=1
KAFKA_NUM_RECOVERY_THREADS_PER_DATA_DIR=1
KAFKA_LOG_RETENTION_HOURS=168
# KAFKA_LOG_SEGMENT_BYTES=1073741824
# KAFKA_LOG_RETENTION_CHECK_INTERVAL_MS=300000

```

I have commented out the block for `SSL`, and I'm trying to run the cluster with `PLAINTEXT` protocol.

The cluster starts properly, I have tested the configuration with `kafkacat`. But when I try to use `kafka_ex` it's not connecting to the cluster. I am always getting the following error.

```
15:24:46.325 [warn]  Ignoring ssl_options [cacertfile: "/home/arjun/work/madsiot/kafka_ex-master/ssl/ca-cert", certfile: "/home/arjun/work/madsiot/kafka_ex-master/ssl/cert.pem", keyfile: "/home/arjun/work/madsiot/kafka_ex-master/ssl/key.pem"] because use_ssl is false.  If you do not intend to use ssl and want to remove this warning, set `ssl_options: []` in the KafkaEx config.
 
15:24:46.331 [debug] Successfully connected to broker "localhost":9093
 
15:24:46.332 [debug] Successfully connected to broker "localhost":9094
 
15:24:46.332 [debug] Successfully connected to broker "localhost":9095
 
15:24:46.739 [info]  Application kafka_ex exited: KafkaEx.start(:normal, []) returned an error: an exception was raised:
    ** (RuntimeError) Brokers sockets are closed
        (kafka_ex) lib/kafka_ex/server_0_p_10_and_later.ex:106: KafkaEx.Server0P10AndLater.kafka_server_init/1
        (stdlib) gen_server.erl:374: :gen_server.init_it/2
        (stdlib) gen_server.erl:342: :gen_server.init_it/6
        (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
** (Mix) Could not start application kafka_ex: KafkaEx.start(:normal, []) returned an error: an exception was raised:
    ** (RuntimeError) Brokers sockets are closed
        (kafka_ex) lib/kafka_ex/server_0_p_10_and_later.ex:106: KafkaEx.Server0P10AndLater.kafka_server_init/1
        (stdlib) gen_server.erl:374: :gen_server.init_it/2
        (stdlib) gen_server.erl:342: :gen_server.init_it/6
        (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

I have disabled the `SSL` option in `config.exs` as I wanted to try out with PLAINTEXT first. 

I am not able to understand what configuration I am doing incorrectly. Some help is appreciated.

**Comment by joshuawscott** [comment]
I tried 2.6 with some of the kafkaex using apps we have, and it works fine. This appears to be an issue with the way the docker-compose file is set up.

The current master isn't compatible because I believe the KAFKA_ADVERTISED_PORT is deprecated, for example.

For this file you're presenting, we need another environment variable:
```
KAFKA_INTER_BROKER_LISTENER_NAME=INSIDE
```
This gets me farther, although I see SSL errors in the logs. Checking around it looks like kafka is validating the hostname from version 2.0+. I added this to ignore that check:
```
KAFKA_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM=
```
This allows kafka_ex to start up, but the docker containers are logging 'waiting for kafka to be ready', presumably from:
https://github.com/wurstmeister/kafka-docker/blob/master/create-topics.sh#L15

**Comment by arjun289** [comment]
@joshuawscott 
I have already used the env variable, if you check the docker env file
```
KAFKA_INTER_BROKER_LISTENER_NAME=INSIDE
```
I have tried this one too, setting
```
KAFKA_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM=
```

You are right about 
```
docker containers are logging 'waiting for kafka to be ready', presumably from:
https://github.com/wurstmeister/kafka-docker/blob/master/create-topics.sh#L15
```
I have been getting the same logs as above, `waiting for kafka to be ready`, when I run the cluster.

Is there some project you can give link for where you have used this? It would be helpful if you can guide me towards that.

**Comment by joshuawscott** [comment]
To be clear, it seems most of the tests pass, with the exception of a few that are testing things like available versions that differ between 0.11 and 2.6.
Working config.exs has:
```
config :kafka_ex,
[
    brokers: [{"host1", 9092}, {"host2", 9092}, {"host3", 9092}],
    consumer_group: :no_consumer_group,
    sync_timeout: 20000,
    use_ssl: false,
    kafka_version: "kayrock"]

```

**Comment by arjun289** [comment]
Thanks, @joshuawscott I managed to make it work.
